### DISCUSSION TITLE
What does "0 hops" mean?

### DISCUSSION THREAD

**Discussion by codegrue** [pschatzmann/arduino-audio-tools#1514] [Q&A] [answered]
I am playing WAV files off SPIFFS and after a few plays, I stop getting sound although the loop is happening and copier.copy() is getting called. The output shows something like this repeating indefinitely. The sounds seems to stop when it goes from "1 hops" to "0 hops":

```
...
[I] StreamCopy.h : 149 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 149 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 149 - StreamCopy::copy  24 -> 24 -> 24 bytes - in 1 hops
[I] StreamCopy.h : 149 - StreamCopy::copy  0 -> 0 -> 0 bytes - in 0 hops
[I] StreamCopy.h : 149 - StreamCopy::copy  0 -> 0 -> 0 bytes - in 0 hops
[I] StreamCopy.h : 149 - StreamCopy::copy  0 -> 0 -> 0 bytes - in 0 hops
... (repeats forever)
```

If I output the position and size of the WAV file, I can see that it never reaches the end of the file. wavFile.available() also shows as true when in this state so the wav file seems to not be "completing."

I'm not sure this is a bug or something I am doing wrong.  I have tried the following code to "reset" the wav file between playthoughs.

```
        copier.end();
        wavFile.flush();
        wavFile.close();
        encoder.end();
```

**Comment by pschatzmann** [comment, marked as answer]
available() provides the number of bytes: what value do you get ?
I really expect that copy returns 0 bytes copied, when you are at the end...

**Reply by codegrue** [reply]
I changed my copy to this and a think it may have fixed it as a workaround... still testing:

```
int size = copier.copy();
if (size == 0)
    wavFile.seek(wavFile.size()); // force moving to the end
```

**Reply by codegrue** [reply]
It seems available() gives a number of bytes remaining, but copy() gives 0 so this repeats infinitely. In my tests, when the problem occurs these are the values:

wavFile.available() = 2
wavFile.position() = 11288
wavFile.size() = 11290
copier.copy() = 0

**Reply by pschatzmann** [reply]
The copier contains some logic to make sure that only full frames are processed. In the latest version you can deactivate this, by setting the mininum copy size to 1: with setMinCopySize(1).

This should resolve your issue.

**Reply by codegrue** [reply]
Thanks for the fix. But if it cannot complete the frame, should it also adjust the file to be "complete"?

**Reply by pschatzmann** [reply]
Do whatever is more convenient to you: I guess you would not hear any difference since the end of the audio usually just contains 0 values from fading out.

A wave file starts with a header of 44 bytes and if it is stereo 16 bits, a full frame of left and right data has 4 bytes: so your file seems to end with a left sample and the corresponding right sample is missing.

**Reply by codegrue** [reply]
Thank you for your help and this amazing library! The work you have put into this is very appreciated.

**Reply by z-l-p** [reply]
I thought you'd like to know that this solved a mysterious problem on my internet radio project: After upgrading the library, several internet radio streams would stop after a few seconds (with the same errors reported by codegrue). 
setMinCopySize(1) on my copier restored everything to normal.

**Reply by pschatzmann** [reply]
Thanks for the feed-back: I changed the default setting to be 1

**Reply by z-l-p** [reply]
Confirmed that the new default in commit fb80700 works in my application. Thanks!
